app/models/authenticator/httpd.rb

```diff
@@ -130,22 +138,37 @@ def user_details_from_external_directory(username)
     end
 
     def user_details_from_headers(username, request)
+      user_headers = user_headers_from_request(request)
+
       if debug_auth?
         log_auth_debug("user_details_from_headers(username=#{username})")
-
-        remote_user_headers = %w[X-REMOTE-USER X-REMOTE-USER-FIRSTNAME X-REMOTE-USER-LASTNAME X-REMOTE-USER-FULLNAME X-REMOTE-USER-EMAIL X-REMOTE-USER-DOMAIN X-REMOTE-USER-GROUPS]
-        logged_headers = remote_user_headers.map { |rh| "  %-24{key} = \"%{val}\"" % {:key => rh, :val => request.headers[rh]} }
-
         log_auth_debug("External-Auth remote user request.headers:")
-        log_auth_debug(logged_headers)
+        user_headers.each { |k, v| log_auth_debug("  %-24{key} = \"%{val}\"" % {:key => k, :val => v}) }
       end
+
+      groups     = CGI.unescape(user_headers['X-REMOTE-USER-GROUPS'] || '').split(/[;:,]/)
       user_attrs = {:username  => username,
-                    :fullname  => request.headers['X-REMOTE-USER-FULLNAME'],
-                    :firstname => request.headers['X-REMOTE-USER-FIRSTNAME'],
-                    :lastname  => request.headers['X-REMOTE-USER-LASTNAME'],
-                    :email     => request.headers['X-REMOTE-USER-EMAIL'],
-                    :domain    => request.headers['X-REMOTE-USER-DOMAIN']}
-      [user_attrs, (CGI.unescape(request.headers['X-REMOTE-USER-GROUPS'] || '')).split(/[;:,]/)]
+                    :fullname  => user_headers['X-REMOTE-USER-FULLNAME'],
+                    :firstname => user_headers['X-REMOTE-USER-FIRSTNAME'],
+                    :lastname  => user_headers['X-REMOTE-USER-LASTNAME'],
+                    :email     => user_headers['X-REMOTE-USER-EMAIL'],
+                    :domain    => user_headers['X-REMOTE-USER-DOMAIN']}
+
+      [user_attrs, groups]
+    end
+
+    private def user_headers_from_request(request)
+      %w[
+        X-REMOTE-USER
+        X-REMOTE-USER-FULLNAME
+        X-REMOTE-USER-FIRSTNAME
+        X-REMOTE-USER-LASTNAME
+        X-REMOTE-USER-EMAIL
+        X-REMOTE-USER-DOMAIN
+        X-REMOTE-USER-GROUPS
+      ].each_with_object({}) do |k, h|
+        h[k] = request.headers[k]&.force_encoding("UTF-8")
```

Comment:
I'm very repetitive... should we return a copy of the string so we don't modify the one in the header in place?